src/profiler.ts

```diff
@@ -13,57 +13,236 @@
+ */
+export class Profiler extends common.ServiceObject {
+  private config: ProfilerConfig;
+  private logger: Logger;
+  private profileTypes: string[];
+  private profilingTimeoutId: undefined|NodeJS.Timer;
+
+  // Public for testing.
+  timeProfiler: TimeProfiler|undefined;
 
   constructor(config: ProfilerConfig) {
-    this.config = config;
+    config = common.util.normalizeArguments(null, config);
+    const serviceConfig = {
+      baseUrl: API,
+      scopes: ['https://www.googleapis.com/auth/monitoring.write'],
+      packageJson: pjson,
+    };
+    super({parent: new common.Service(serviceConfig, config), baseUrl: '/'});
 
-    this.service = new common.Service(
-        {
-          baseUrl: 'https://cloudprofiler.googleapis.com/v2',
-          scopes: ['https://www.googleapis.com/auth/monitoring.write'],
-        },
-        this.config);
+    this.config = config;
 
     this.logger = new common.logger({
       level: common.logger.LEVELS[config.logLevel as number],
       tag: pjson.name
     });
+
+    // TODO: enable heap profiling once heap-profiler implemented.
+    this.profileTypes = [];
+    if (!this.config.disableTime) {
+      this.profileTypes.push(ProfileTypes.Wall);
+      this.timeProfiler =
+          new TimeProfiler(this.config.timeSamplingIntervalMicros);
+    }
   }
 
-  // Begins collection and uploading of profiles.
-  // If profiling fails or another problem is encountered, the returned promise
-  // will be rejected.
-  // TODO: implement
-  // TODO: explain failure situations.
+  /**
+   * Starts and endless loop to poll Stackdriver Profiler server for
+   * instructions, and collects and uploads profiles as requested.
+   * If there is a problem when collecting a profile or uploading a profile to
+   * Stackdriver Profiler, this problem will be logged at the debug level.
+   * If there is a problem polling Stackdriver Profiler for instructions
+   * on the type of profile created, an error will be thrown.
+   */
   async start(): Promise<void> {
-    return Promise.reject(
-        new Error('start() is unimplemented for ProfileAgent'));
+    return this.createAndUploadProfile(0);
+  }
+
+
+  /**
+   * Talks to Stackdriver Profiler server to create profile when instructed
+   * and then upload that profile.
+   *
+   * @param delayMillis - milliseconds to wait before polling Stackdriver
+   * Profiler server to see if it's time to create a profile and profile.
+   */
+  async createAndUploadProfile(delayMillis: number): Promise<void> {
+    this.profilingTimeoutId = setTimeout(async () => {
+      const startCreateMillis = Date.now();
+      const prof = await this.createProfile();
+      await this.profileAndUpload(prof);
+      const endCreateMillis = Date.now();
+      this.createAndUploadProfile(
+          this.config.minTimeBetweenProfilesMillis -
+          (endCreateMillis - startCreateMillis));
+    }, delayMillis);
+    this.profilingTimeoutId.unref();
+  }
```

Comment:
An alternative way to write this function would be:

```js
async createAndUploadProfile() {
  const startCreateMillis = Date.now();
  prof = await this.createProfile();
  await this.profileAndUpload(prof);
  const endCreateMillis = Date.now();

  // Schedule the next profile.
  const delay = this.config.minTimeBetweenProfilesMillis - (endCreateMillis - startCreateMillis);
  setTimeout(this.createAndUploadProfile.bind(this), delay).unref();
}
```

But don't change code to this just yet.

Other things to think about:
* createProfile and profileAndUpload are called back to back. Do they need to be separate functions?
* While setTimeout correctly deals with negative timeouts, it would be good to clamp the value at 0 on the low end.
* This function is creating, uploading and scheduling, but the name only describes the first two verbs. Also note that start is a single line function that does a single opeartion: 'create and upload profile', but in reality it is running and async loop.
* Another way to fix this might be to split this function into two logical parts: the scheduling logic and the profiling logic. If you decide to follow this, perhaps your original structure of the could is better compared to the inverted form I gave earlier in this comment.